audiomentations/core/composition.py

```diff
@@ -106,25 +107,104 @@ def __call__(self, magnitude_spectrogram):
         return magnitude_spectrogram
 
 
+class SomeOf(BaseCompose):
+    """
+    SomeOf randomly picks several (one or plus) of the given transforms when called, and applies these
+    transforms.
+
+    Example usage:
+
+    ```
+    augment = SomeOf([
+        TimeStretch(min_rate=0.8, max_rate=1.25, p=1.0),
+        PitchShift(min_semitones=-4, max_semitones=4, p=1.0),
+        Gain(min_gain_in_db=-12, max_gain_in_db=-6, p=1.0),
+    ])
+    # Generate 2 seconds of dummy audio for the sake of example
+    samples = np.random.uniform(low=-0.2, high=0.2, size=(32000,)).astype(np.float32)
+    # Augment/transform/perturb the audio data
+    augmented_samples = augment(samples=samples, sample_rate=16000)
+    # Result: Several of the three transforms above were applied. The minimum number of applied transforms is 1
+    and the maximum is 3. It is not possible that no transform is applied.
+    ```
+
+    """
+
+    def __init__(self, transforms, p: float = 1.0):
+        super().__init__(transforms, p)
+        self.transform_indexes = []
+        self.nbr_transforms_to_apply = None
+        self.should_apply = True
+
+    def randomize_parameters(self, *args, **kwargs):
+        super().randomize_parameters(*args, **kwargs)
+        self.should_apply = random.random() < self.p
+        if self.should_apply:
+            self.nbr_transforms_to_apply = random.randint(1, len(self.transforms))
+            all_indexes_transforms = list(np.arange(len(self.transforms)))
+            print(all_indexes_transforms)
+            self.transform_indexes = sorted(
+                random.sample(all_indexes_transforms, self.nbr_transforms_to_apply)
+            )
+        return self.transform_indexes
+
+    def __call__(self, *args, **kwargs):
+        if not self.are_parameters_frozen:
+            kwargs["apply_to_children"] = False
+            self.randomize_parameters(*args, **kwargs)
+
+        if self.should_apply:
+            if "apply_to_children" in kwargs:
+                del kwargs["apply_to_children"]
+            print(self.transform_indexes[0])
+            transformed_data = self.transforms[self.transform_indexes[0]](
```

Comment:
Hmm, transformed_data can be either samples or magnitude_spectrogram.

If it is samples, that means that the next transform is also a waveform transform, which means that it expects (samples, sample_rate) as input.

Perhaps it's easier/better to write the for-loop for waveform and spectrogram independently?

And maybe check inside \_\_init\_\_ whether the given transforms are for waveform or spectrogram?

Reply:
Yes, you're right, I haven't thought about it. :) I'll modify it as well !

Reply:
Thanks :)